test/rules/null_check_on_nullable_type_parameter_test.dart

```diff
@@ -18,13 +18,49 @@ class NullCheckOnNullableTypeParameterTestLanguage300 extends LintRuleTest
   @override
   String get lintRule => 'null_check_on_nullable_type_parameter';
 
-  test_nullAssertPattern() async {
+  test_nullAssertPattern_list() async {
     await assertDiagnostics(r'''
-void f<T>((T?, T?) p){
+f<T>(List<T?> l){
+  var [x!, y] = l;
+}
+''', [
+      lint(26, 1),
+    ]);
+  }
+
+  test_nullAssertPattern_map() async {
+    await assertDiagnostics(r'''
+f<T>(Map<String, T?> m){
+  var {'x': y!} = m;
+}
+''', [
+      lint(38, 1),
+    ]);
+  }
+
+  @FailingTest(issue: 'https://github.com/dart-lang/linter/issues/4218')
+  test_nullAssertPattern_object() async {
```

Comment:
I suspect that `matchedValueType`, for a field in an object pattern, is returning the type of the field being matched. In the test, that will be `Object?` and not a type parameter type. I think I steered you wrong, and you really do need for `A` to be parameterized (`class A<E>`), and that you need to pass in the nullable type parameter type in the argument list (`A<T?> u`).

But I think you don't need the parameter `T? t` at all because that `t` isn't referenced. I think the `t` in `A(a: t!)` is a new local variable that shadows the parameter.

Reply:
Ah, yes. Thanks!